src/solvers/dg_unstructured_quad/dg_2d.jl

```diff
@@ -369,3 +307,33 @@ function calc_surface_integral!(du, mesh::UnstructuredQuadMesh,
 
   return nothing
 end
+
+
+# This routine computes the maximum value of the discrete metric identities necessary to ensure
+# that the approxmiation will be free-stream preserving (i.e. a constant solution remains constant)
+# on a curvilinear mesh.
+#   Note! Independent of the equation system and is only a check on the discrete mapping terms.
+#         Can be used for a metric identities check on CurvedMesh{2} or UnstructuredQuadMesh
+function calc_discrete_metric_identities(dg::DGSEM, cache)
+  @unpack derivative_matrix = dg.basis
+  @unpack contravariant_vectors = cache.elements
+
+  metric_id_dx = zeros(eltype(contravariant_vectors), nnodes(dg), nnodes(dg))
+  metric_id_dy = zeros(eltype(contravariant_vectors), nnodes(dg), nnodes(dg))
+  metric_ids_1 = zeros(eltype(contravariant_vectors), nnodes(dg), nnodes(dg), nelements(cache.elements))
+  metric_ids_2 = zeros(eltype(contravariant_vectors), nnodes(dg), nnodes(dg), nelements(cache.elements))
+
+  for element in eachelement(dg, cache)
+    # compute D*Ja_1^1 + Ja_1^2*D^T
+    @views mul!(metric_id_dx, derivative_matrix, contravariant_vectors[1, 1, :, :, element])
+    @views mul!(metric_id_dy, contravariant_vectors[1, 2, :, :, element], derivative_matrix')
+    metric_ids_1[:, :, element] .= abs.(metric_id_dx + metric_id_dy)
+
+    # compute D*Ja_2^1 + Ja_2^2*D^T
+    @views mul!(metric_id_dx, derivative_matrix, contravariant_vectors[2, 1, :, :, element])
+    @views mul!(metric_id_dy, contravariant_vectors[2, 2, :, :, element], derivative_matrix')
+    metric_ids_2[:, :, element] .= abs.(metric_id_dx + metric_id_dy)
+  end
+
+  return max( maximum(metric_ids_1), maximum(metric_ids_2) )
+end
```

Comment:
This is probably not relevant for performance, but you don't need to allocate storage for all elements when you only take the maximum at the end. Would it make sense to do the reduction on the fly to avoid many allocations?

Reply:
yeah, that is a good point. Now that this check is working I can streamline it

Reply:
Now this function is much more compact and the way to generalize it to 3D is a little bit easier to see